File: Codes in C++/Array/Trapping Rainwater.cpp
class Solution {
public:
    int trap(vector<int>& arr) {
        int n = arr.size();
        int lMax, rMax, l, r, total;
        total = lMax = rMax = 0;
        l = 0;
        r = n - 1;
        while (l < r) {
            if (arr[l] <= arr[r]) {
                if (lMax > arr[l])
                    total += lMax - arr[l];
                else
                    lMax = arr[l];
                l++;
            } else {
                if (rMax > arr[r])
                    total += rMax - arr[r];
                else
                    rMax = arr[r];
                r--;
            }
        }
        return total;
    }
};

File: Codes in Python3/Two Pointers/Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l,r=0,len(height)-1
        lMax,rMax=height[0],height[len(height)-1]
        maxArea=0

        while(l<r):
            if height[l]<height[r]:
                if lMax<height[l]:
                    lMax=height[l]
                else:
                    maxArea+=lMax-height[l]
                l+=1
            else:
                if rMax<height[r]:
                    rMax=height[r]
                else:
                    maxArea+=rMax-height[r]
                r-=1
        return maxArea

# TC=O(n)
# SC=O(1)